contracts/StorageManager.sol

```diff
@@ -172,177 +156,181 @@ contract StorageManager {
+        }
+
+        if (amount == 0) {
+            amount = maxWithdrawableFunds;
+        }
+
+        require(amount <= maxWithdrawableFunds, "StorageManager: Amount is too big");
+        agreement.availableFunds = agreement.availableFunds.sub(amount);
+        msg.sender.transfer(amount);
+        emit AgreementFundsWithdrawn(agreementReference, amount);
     }
 
     /**
+    >> FOR PROVIDER
     @notice payout already earned funds of one or more Agreement
     @dev 
-    - any safeMath operations in this function don't cause a deadlock, as a possible revert just means we have to add less agreementReferences
-    - Provider must call an expired agreement themselves as soon as the agreement is expired, to add back the capacity.
+    - Provider must call an expired agreement themselves as soon as the agreement is expired, to add back the capacity to their Offer.
     @param agreementReferences reference to one or more Agreement
     */
     function payoutFunds(bytes32[] memory agreementReferences) public {
+        Offer storage offer = offerRegistry[msg.sender];
+        require(offer.availableCapacity != 0, "StorageManager: Offer for this Provider doesn't exist");
         uint256 toTransfer;
-        for(uint8 i = 0; i < agreementReferences.length; i++) {
-            Agreement storage agreement = offerRegistry[msg.sender].agreementRegistry[agreementReferences[i]];
-            require(agreement.startDate != 0, "StorageManager: Agreement not active");
-            // check if agreement is expired
-            //NO_OVERFLOW REASONING: see REF_DURATION
-            if((agreement.startDate + agreement.numberOfPeriodsDeposited) * agreement.billingPeriod < now) {
-                //NO_OVERFLOW reasoning: numberOfPeriodsWithdrawn is always less than or equal to numberOfPeriodsDeposited and REF_MAX_TRANSFER
-                toTransfer = toTransfer.add((agreement.numberOfPeriodsDeposited - agreement.numberOfPeriodsWithdrawn) * agreement.billingPrice);
-                //reset agreement to clean state
-                agreement.numberOfPeriodsWithdrawn = 0;
-                agreement.numberOfPeriodsDeposited = 0;
-                agreement.startDate = 0;
-                // check if Offer is still active
-                if(offerRegistry[msg.sender].availableCapacity != 0) {
+        for (uint8 i = 0; i < agreementReferences.length; i++) {
+            Agreement storage agreement = offer.agreementRegistry[agreementReferences[i]];
+            require(agreement.size != 0, "StorageManager: Agreement for this Offer doesn't exist");
+
+            // TODO: [Q] Should we disallow to payout from inactive agreement? Can it cause some deadlock?
+            require(agreement.lastPayoutDate != 0, "StorageManager: Agreement is inactive");
+
+            uint256 spentFunds = _calculateSpentFunds(agreement);
+            agreement.availableFunds = agreement.availableFunds.sub(spentFunds);
+            toTransfer = toTransfer.add(spentFunds);
```

Comment:
toTransfer is 0 at this point, why don't you just assign `toTransfer = spentFunds`? Or better why don't we use `spentFunds` directly?

Reply:
Because it is in loop in order to support paying out multiple Agreements in one call 😉 So the loop calculates `toTransfer` and only then transfers the total sum.